src/main/java/org/codehaus/groovy/runtime/MetaClassHelper.java

```diff
@@ -1016,4 +1018,21 @@ public static String convertPropertyName(String prop) {
         }
         return BeanUtils.decapitalize(prop);
     }
+
+    // GROOVY-9075
+    public static void checkMetaClassProperty(MetaClass oldValue, Object newValue) {
+        if (null == newValue) return;
+
+        if (!(newValue instanceof MetaClass)) {
+            throw new IllegalArgumentException("the new metaClass[" + newValue.getClass() + "] is not an instance of " + MetaClass.class);
+        }
+
+        MetaClass newMetaClass = (MetaClass) newValue;
+        if (!newMetaClass.getTheClass().isAssignableFrom(oldValue.getTheClass())) {
```

Comment:
I am unsure that this is in fact the real requirement. The following currently works:
```
class MyMeta extends MetaClassImpl {
    MyMeta(Class klass) { super(klass) }
}
class S2 extends MetaClassImpl {
    S2(Class klass) { super(klass) }
}

def s1 = "foo${1+1}"
def m1 = new MyMeta(GString)
m1.initialize()
s1.metaClass = m1

def s2 = 'bar'
def m2 = new MyMeta(String)
m2.initialize()
s2.metaClass = m2

println m1.theClass.isAssignableFrom(m2.theClass)
println m2.theClass.isAssignableFrom(m1.theClass)
s1.metaClass = s2.metaClass
```
But wouldn't that now fail?

Reply:
It's a breaking change... but I am not sure whether it will break real use cases...